Add SphericalCompression CoordMap

geoffrey4444 opened the pull request:
## Proposed changes

Add a time-dependent CoordMap that implements a "size control" style map, equivalent to using DistortedSphereMapFunctionSphere in SpEC. The map compresses a spherical region, with the amount of compression going to zero at some max radius and leveling off to a constant value at some min radius.

As far as I can tell, this is the first CoordMap whose behavior depends on the region. Specifically, while the map is continuous, the Jacobian is not.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
There is not yet a use for the map, so there is nothing to upgrade. This PR is the first step toward implementing a TimeDependence that can track the growth of the black holes in a binary as they move closer together.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 committed:
Add SphericalCompression CoordMap

geoffrey4444 commented:
The one failed test gives an error unrelated to changes in this PR:

53/64 Test #1192: Integration.Parallel.AlgorithmParallelWithBalancing ................***Failed  Required regular expression not found. Regex=[Objects migrating: 14
114
]  0.00 sec
115
/usr/bin/sh: 1: /work/build/bin/Test_AlgorithmParallel: not found

This is ready for a look @markscheel whenever you get a chance

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,294 @@
+ * \rho_{\rm max}\f$, where instead the map applies compression that is
+ * spherically symmetric about the center \f$C^i\f$. The amount of
+ * compression decreases linearly from a maximum at
+ * \f$\rho \leq \rho_{\rm min}\f$ to zero at \f$\rho = \rho_{\rm max}\f$.
```
`\rho = \rho_{\rm min}`

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,294 @@
+ * \frac{\rho_{\rm max} / \rho - 1}{\rho_{\rm max} - \rho_{\rm min}} \rho^i,
+ * \mbox{ } \rho_{\rm min} \leq \rho < \rho_{\rm max}, \\
+ * x^i &= \xi^i, \mbox{ } \rho_{\rm max} < \rho,
+ * \f}
```
Readability:  In normal LaTeX one can do different cases with a single equals sign and a big opening curly brace, as it is usually written in papers, using 
```
x^i = \left\{\begin{array}{ll}
\xi^i - \frac{\lambda_{00}(t)}{\sqrt{4\pi}} \frac{\rho^i}{\rho_{\rm min}}, & \rho < \rho_{\rm min},\\
\xi^i, & \rho > \rho_{\rm max}
\end{array}\right.
```
I would think this would be possible in doxygen latex.

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,294 @@
+ * \rho &= \frac{r+\frac{\lambda_{00}(t)}{\sqrt{4\pi}}\frac{\rho_{\rm
+ * max}}{\rho_{\rm max}-\rho_{\rm min}}}{1 + \frac{\lambda_{00}(t)}{\sqrt{4\pi}}
+ * \frac{1}{\rho_{\rm max} - \rho_{\rm min}}}.
+ * \f}
```
Readability: can you avoid fracs of fracs here and below?  It makes things really hard to read. One way is to write `(numerator) (denominator)^{-1}` instead of `\frac{numerator}{denominator}`.  Or define auxiliary variables.

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,294 @@
+
+namespace domain::CoordinateMaps::TimeDependent {
+/// \cond HIDDEN_SYMBOLS
+template <size_t Dim>
```
Do we want to remove the template on Dim, since we will only care about this for Dim=3?

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,385 @@
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> SphericalCompression<3>::operator()(
+    const std::array<T, 3>& source_coords, const double time,
```
`using ResultType = tt::remove_cvref_wrap_t<T>` would (slightly) shorten the code.

markscheel replied:
I actually meant use a local type alias `ResultType` inside those functions in which `tt::remove_cvref_wrap_t<T>` was repeated frequently, but I'm ok with a global type alias in an unnamed namespace.

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,385 @@
+            make_not_null(&(*it_result)), *it_radial_coord, *it_radius,
+            lambda_y, min_radius_, max_radius_);
+      }
+    }
```
Looks to me like `correct_mapped_coordinate_or_frame_velocity` should work when passed in DataVectors as its first three arguments.  So then it shouldn't be necessary to iterate over values in the container, or to have a separate case here for double.

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,385 @@
+            make_not_null(&(*it_result)), *it_radial_coord, *it_radius,
+            lambda_y, min_radius_, max_radius_);
+      }
+    }
```
Like above, can this entire section be replaced by the following?
```
      correct_mapped_coordinate_or_frame_velocity(
          make_not_null(&gsl::at(result, i)), gsl::at(source_rad_position, i),
          source_radius, dt_lambda00_y00(f_of_t_name_, time, functions_of_time),
          min_radius_, max_radius_);
```

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,385 @@
+                             source_point_rad_position, *it_radius, lambda_y,
+                             min_radius_, max_radius_, i, j);
+        }
+      }
```
Ditto for not needing loop or separate case for T=double.

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,294 @@
+ *
+ * ### Inverse map
+ *
+ * The map is invertible if \f$\rho_{\rm min} - \rho_{\rm max} < \lambda_{00}(t)
+ * / \sqrt{4\pi} < \rho_{\rm min}\f$. In this case, the
```
I think this condition is not exactly invertibility.  If this condition is violated, then `r` is negative in Eq 19 or 20, which maybe can be dealt with by careful consideration of signs, but it's easier to just avoid that regime.  The map is singular for both 'equals' cases of this condition, so requiring this condition ensures that for lambda=0 and all lambda that obey this condition the map remains nonsingular.

Edit: let me add that when this condition is violated, the map doesn't do what one thinks it does anymore:  a larger lambda will result in an expansion (and a coordinate inversion because the radius is negative) instead of a compression.

geoffrey4444 replied:
It sounds like you're saying the condition is sufficient but not necessary for invertibility? If so, I can flesh out comments along the lines you suggest

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,385 @@
+                             (max_radius - min_radius));
+      }
+    } else {
+      if (i == j) {
+        *input = 1.0;
+      } else {
+        *input = 0.0;
+      }
+    }
```
Do we want to Error in this case?  A Block with rho >= rho_max shouldn't even contain a SphericalCompression map.

markscheel reviewed (commented):
Various implementation and docs comments.

Global question: should this map actually be two different maps:
  - One for rho <= rho_min
  - One for rho_min <= rho <= rho_max

Then the maps will Error if evaluated outside their range.  (The 'two maps' could perhaps be one class with a template parameter denoting the range).  The reason I like two maps is that in actual usage we will require that rho_min and rho_max be on Block boundaries.  So the easiest thing to do would be for Blocks with rho<= rho_min to have one map, Blocks with rho_min<=rho<=rho_max to have another map, and blocks with rho>=rho_max to not have any compression map at all.

geoffrey4444 commented:
> Various implementation and docs comments.
> 
> Global question: should this map actually be two different maps:
> 
> * One for rho <= rho_min
> * One for rho_min <= rho <= rho_max
> 
> Then the maps will Error if evaluated outside their range. (The 'two maps' could perhaps be one class with a template parameter denoting the range). The reason I like two maps is that in actual usage we will require that rho_min and rho_max be on Block boundaries. So the easiest thing to do would be for Blocks with rho<= rho_min to have one map, Blocks with rho_min<=rho<=rho_max to have another map, and blocks with rho>=rho_max to not have any compression map at all.

In my next fixup, I will implement this change: a template parameter will set which region you're in (inner or middle), and blocks in the outer region, where the transformation should be the identity, I just won't apply any map when I implement the time dependence

geoffrey4444 commented:
Thanks @markscheel for the feedback! I pushed multiple fixup commits that group my responses to the different suggestions you made, and I also rebased on the latest develop

geoffrey4444 commented:
Tests are fixed...a little mistake before my most recent fixup made one of 
Y tests always fail

markscheel commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,372 @@
+        radial_position(target_coords, center_)};
+    const double target_radius{magnitude(target_radial_position)};
+    std::array<double, 3> result{target_radial_position};
+    double scale{};
```
Please define `scale` and `result` after the checks, i.e. just before they are used.

markscheel commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+#include "Utilities/StdArrayHelpers.hpp"
+
+namespace {
+// Generates the map, time, an a FunctionOfTime
```
`an` -> `and`

markscheel commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+          initial_time, initial_coefficients, expiration_time);
+}
+
+// Generates the map, time, an a FunctionOfTime, but hides internal details (min
```
an -> and

markscheel commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+  const double theta{theta_dis(*generator)};
+  const double phi{phi_dis(*generator)};
+
+  // Chose a point of unit radius with the randomly selected theta, phi.
```
Chose -> Choose

markscheel reviewed (commented):
A few minor things; please squash!

geoffrey4444 commented:
Thanks again for the review, @markscheel ! I went ahead and squashed. This is ready for a second reviewer to take a look

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,322 @@
+ * where \f$\rho^i = \xi^i - C^i\f$ is the Euclidean radial position vector in
+ * the unmapped coordinates with respect to the center \f$C^i\f$, \f$\rho =
+ * \sqrt{\delta_{kl}\left(\xi^k - C^l\right)\left(\xi^l - C^l\right)}\f$ is the
+ * Euclidean magnitude of \f$\rho^i\f$, and \f$\rho^j = \delta_{ij} \rho^i\f$.
```
The final inline equation probably intended a lower index `\rho_j`.

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,322 @@
+ *
+ * ### Inverse map
+ *
+ * For \f$\lambda_{00}(t)\f$ that satify
```
`satify`->`satisfy`

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,372 @@
+// This function computes the (i,j) component of the Jacobian matrix
+// \f$\partial x^i / \partial \xi^j.\f$
+template <bool InteriorMap, typename T>
+void jacobian_component(gsl::not_null<ResultType<T>*> input,
```
This function is reasonably short, and used only in one place. Perhaps consider just putting its contents in the calling function instead, unless there's a clear gain of factoring it out that I'm missing.

moxcodes replied:
In the fixup, you've put the contents of the function inline, but forgot to delete the helper function in the anonymous namespace.

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,322 @@
+          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+          functions_of_time) const noexcept;
+
+  std::optional<std::array<double, 3>> inverse(
```
Why is the inverse map only callable with a `double`?

geoffrey4444 replied:
I can't recall, but I generalized it to support doubles or DataVectors

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,372 @@
+        radial_position(target_coords, center_)};
+    const double target_radius{magnitude(target_radial_position)};
+    if constexpr (InteriorMap) {
+      if (!(target_radius <= min_radius_ - lambda_y)) {
```
prefer non-inverted logic if it's viable -- here it seems like it could be `target_radius > min_radius_ - lambda_y`. Similar suggestion for the next check.

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,372 @@
+      scale = 1.0 / (1.0 - lambda_y / min_radius_);
+    } else {
+      scale = (1.0 + lambda_y * max_radius_ / (max_radius_ - min_radius_) /
+                         target_radius) /
```
this could be ` / (target_radius * (max_radius_ - min_radius_))` which would be fewer divisions, I think -- or, better yet, it seems like the `(max_radius_ - min_radius_)` term could be factored into the first term of both the numerator and denominator.

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,372 @@
+    const std::unordered_map<
+        std::string, std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+        functions_of_time) const noexcept {
+  // Obtain the Jacobian and then compute its inverse numerically
```
This can probably be optimized without much mathematical trouble. The forward jacobian is either a factor times the Kronecker delta or of the form (a * \delta^i{}_j + b n^i n_j), for a normal vector n. In the latter case, I think the inverse is something like (1/a \delta^i{}_j - b/(a (b + a))) n^i n_j).
I won't insist that you implement this for this PR if it seems inconvenient, but perhaps include a code comment that this is a place where there's a clear opportunity for optimization.

geoffrey4444 replied:
I'll add the comment; while I think you're probably right, I already thought a bit about this, and it seemed inconvenient enough that I went the numerical route. Whoever implemented this in spec came to the same conclusion, not that spec can justify anything for spectre :)

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,372 @@
+#undef DTYPE
+#undef INSTANTIATE
+
+#define INTERIOR_MAP(data) BOOST_PP_TUPLE_ELEM(0, data)
```
This is the same as the previous `INTERIOR_MAP`, so you can omit the first `#undef` and this re-declaration.

moxcodes commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+    "Unit.Domain.CoordinateMaps.SphericalCompression.MissingFofT",
+    "[Domain][Unit]") {
+  ASSERTION_TEST();
+  MAKE_GENERATOR(gen);
```
Most of our assertion tests follow the pattern of wrapping their contents in a `#ifdef SPECTRE_DEBUG ... #endif`, and having an `ERROR("Failed to trigger ASSERT in an assertion test)`, I think to avoid the need for the `std::terminate()` at the end -- is there a reason for preferring to not use that pattern here?
(if you change this assertion test, then also propagate a similar change to the previous assertion test.)

geoffrey4444 replied:
I also fixed the ERROR tests similarly, removing std::terminate() and instead having a "failed to trigger ERROR" call to ERROR() at the end

moxcodes commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+    CHECK_FALSE(map_to_test != map_to_test);
+
+    test_coordinate_map_argument_types(map_to_test, random_point, time,
+                                       functions_of_time);
```
is it usable to check the frame velocity using `test_frame_velocity`?

moxcodes commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+      make_not_null(&functions_of_time), make_not_null(&min_radius),
+      make_not_null(&max_radius), make_not_null(&center), generator);
+
+  const double theta{theta_dis(*generator)};
```
[optional] this code used to create a random point could be factored into another small helper function. It's used three separate times across the tests.

moxcodes commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+        const std::array<double, 3> mapped_point_2{
+            map(orig_point_2, time, functions_of_time)};
+        const std::array<double, 3> scale_factor_1{
+            {(orig_point_1[0] - mapped_point_1[0]) /
```
this test feels like it could be nicer to read as a loop rather than explicit indices 0-2 in as many places as it is.

moxcodes commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,595 @@
+  }
+
+  // Check that a point is invertible with the supplied functions_of_time
+  CHECK(static_cast<bool>(
```
for `std::optional`, we can use `.has_value()` which is more readable (and below)

moxcodes reviewed (commented):
You can squash the fixup along with this last minor note.

moxcodes reviewed (dismissed):
Looks good

markscheel commented:
Wait, why is the inverse map now callable with a DataVector?  All the inverse maps in every CoordinateMap should be callable only with doubles, because the inverse may fail if called for a point out of range (hence the std::optional wrapping the return type).   (If the inverse map is callable with a DataVector, what should it return when it fails for some points but succeeds for other points?  And what should the calling code do in that case?)

moxcodes commented:
Ah, thanks @markscheel -- I asked about it in my review, and @geoffrey4444 didn't know the reason to avoid generalization. That makes good sense.
@geoffrey4444 sorry about that, please revert the inverse map type to only act on `double`s.

markscheel commented:
Maybe every CoordinateMap should have a comment next to `inverse()` saying why it is callable only for doubles.  Because having it called only for doubles can easily be interpreted as a bug, especially since every other function in the CoordinateMap is callable for DataVectors too.

geoffrey4444 committed:
Make CubicScale inverse interface same as other CoordMaps

geoffrey4444 committed:
Add comments explaining why CoordMap inverses only take doubles

geoffrey4444 commented:
Thanks for the feedback, @markscheel and @moxcodes! This is ready for another look when you get a chance.

I just squashed immediately a reversion so the inverse function once again only takes doubles.

I also added a new commit that adds an identical comment to every CoordMap explaining why the inverse only takes doubles.  

Finally, I also added a new commit fixes CubicScale's inverse so that it only takes doubles...for some reason, it was templated on DataType, and then only double and reference to double were instantiated. This is why I was confused: @moxcodes asked me to make the inverse templated on DataType, and when I checked to see if other CoordMaps worked that way, the one I chose to look at was CubicScale, which actually was templated on DataType, even though none of the others are.

moxcodes commented on src/Domain/CoordinateMaps/TimeDependent/CubicScale.cpp:
```diff
@@ -415,25 +411,6 @@ bool operator==(const CubicScale<Dim>& lhs,
 
 #define INSTANTIATE(_, data)                                                 \
   template class CubicScale<DIM(data)>;                                      \
-  template std::optional<                                                    \
```
Don't we still need the instantiations for each `DIM`?

markscheel replied:
I think instantiating the class for each DIM (`template class CubicScale<DIM(data)>;` on new line 413)  is enough to instantiate `inverse()`, since `inverse()` is no longer a member function template.  Whether the tests pass or fail will tell us.

geoffrey4444 replied:
I tried adding such an explicit instantiation:

~~~
 template std::optional<std::array<double, DIM(data)>>                      \
  CubicScale<DIM(data)>::inverse(                                            \
      const std::array<double, DIM(data)>& target_coords, const double time, \
      const std::unordered_map<                                              \
          std::string,                                                       \
          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&         \
          functions_of_time) const noexcept;                                 \
~~~

but I got this compiler error:

~~~
/home/geoffrey/Codes/spectre/spectre/src/Domain/CoordinateMaps/TimeDependent/CubicScale.cpp:424:25: error: duplicate explicit instantiation of 'inverse'
GENERATE_INSTANTIATIONS(INSTANTIATE, (1, 2, 3))
                        ^
/home/geoffrey/Codes/spectre/spectre/src/Domain/CoordinateMaps/TimeDependent/CubicScale.cpp:424:25: note: previous explicit instantiation is here
/home/geoffrey/Codes/spectre/spectre/src/Domain/CoordinateMaps/TimeDependent/CubicScale.cpp:424:25: error: duplicate explicit instantiation of 'inverse'
GENERATE_INSTANTIATIONS(INSTANTIATE, (1, 2, 3))
                        ^
/home/geoffrey/Codes/spectre/spectre/src/Domain/CoordinateMaps/TimeDependent/CubicScale.cpp:424:25: note: previous explicit instantiation is here
/home/geoffrey/Codes/spectre/spectre/src/Domain/CoordinateMaps/TimeDependent/CubicScale.cpp:424:25: error: duplicate explicit instantiation of 'inverse'
GENERATE_INSTANTIATIONS(INSTANTIATE, (1, 2, 3))
~~~

I got no compiler error, and the tests all passed, when I removed this. Could you please clarify why we need it? Do you know why I was getting that compiler error?

moxcodes replied:
Ah, I see -- Mark is correct; the `DIM` is a template parameter of the class, not the function, so we only need the explicit instantiation for the class.

geoffrey4444 commented:
@kidder I wonder if you might take a look at this whenever you get a chance? @markscheel and @moxcodes are happy with it, but it needs a look from a core developer

nilsdeppe commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,322 @@
+ * & \rho_{\rm min} \leq \rho \leq \rho_{\rm max}, \\
+ * 0, & \rho_{\rm max} < \rho.\end{array}\right.
+ * \f}
+ *
```
maybe add `where \lambda^\prime is...`?

nilsdeppe commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.hpp:
```diff
@@ -0,0 +1,322 @@
+ *
+ * ### Inverse Jacobian
+ *
+ * This map finds the inverse Jacobian by first finding the Jacobian and then
```
Question: given the restriction on the inverse map, imposing that restriction here would mean that the Jacobian is (I think) surprisingly simple since it basically boils down to `\xi^i = Constant(t)` for most of them, with just a `1/r^2 \partial_i r` being the only non-trivial factor. Would this be worth doing? Not necessarily now, but even later (filing a feature request)?
@markscheel have you thought about this?

geoffrey4444 replied:
I thought about this briefly but didn't see a straightforward way to get the 1/r^2 \partial_i r term, but then again, my threshold for biting the bullet and doing some algebra is sometimes set too low. If @markscheel has an idea, I'd be happy to open an issue on this with it as a feature request.

geoffrey4444 replied:
I looked again at this, and I think the middle case might be tricky, because it's not obvious to me you can easily solve for \xi^I before taking the derivative, because of the (rho_max/rho - 1)rho^i term

nilsdeppe replied:
Sorry, I'm missing something. So I'm looking at equation (19) in the dox, and these have for `r < rh_min-lamda/sqrt(4pi)` and `rho_max<r` that `\xi~r^i` which means `d\xi^i / dx^j~\delta^i_j(stuff)` where the "stuff" is different in the two cases. In the interior, you have
```
xi^i~r^i(1+A/r)/B +C^i
```
so
```
d xi^i /dx^j~\delta^i_j(1+A/r)/B+r^i(-A/r^2 dr/dx^j)/B
```
That would mean you need to figure out what `dr/dx^j` is, so
```
d r / dx^j = (d/dx^j) sqrt((x^i-C^i)\delta_{ik}(x^k-C^k))
= (1/2r) (d/dx^j) (x^i-C^i)(x^k-C^k) \delta_{ik}
= (1/2r) \delta_{ik} (\delta ^i_j(x^k -C^k) + \delta^k_j(x^i-C^i))
=  (1/2r) (\delta_{kj} (x^k -C^k) + \delta_{kj}(x^k-C^k))
=  (1/r) (x^j -C^j)
```
What I'm not sure about are the lambda factors, since `x^i` is a function of time, so there might be some weirdness there? And maybe I've made some other assumption or simplification that I didn't understand from your comments.

geoffrey4444 replied:
You might be right, but as you said, let's leave this as an issue for a future update

nilsdeppe replied:
Can you please add to the documentation what we think the inverse Jacobian is and explain that it needs to be double checked? A github comment on a PR isn't really a great reminder. Edit: or a GitHub issue is fine too. Just something other than a comment that is basically lost as soon as we merge :)

I also still don't understand your statement about the rho terms. Where is the `rho^i` that I missed?

geoffrey4444 replied:
Let me think carefully about your question about the rho terms I thought you missed and then get back to you in the issue I just opened about this: #2989

nilsdeppe replied:
:+1: thanks!

nilsdeppe commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,357 @@
+// /sqrt(4*M_PI)).
+template <bool InteriorMap, typename T>
+void correct_mapped_coordinate_or_frame_velocity(
+    gsl::not_null<ResultType<T>*> input, const T& source_radial_coord,
```
`const gsl::not_null`

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+namespace {
+// Generate a point in a random direction at a given radius
+auto points_in_random_direction_generator(
+    gsl::not_null<std::mt19937*> generator,
```
`const gsl`

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+// Generates the map, time, and a FunctionOfTime
+template <bool InteriorMap>
+void generate_map_time_and_f_of_time(
+    gsl::not_null<domain::CoordinateMaps::TimeDependent::SphericalCompression<
```
`const gsl` several times

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+// (min radius, max radius, and center) when not needed
+template <bool InteriorMap>
+void generate_map_time_and_f_of_time(
+    gsl::not_null<domain::CoordinateMaps::TimeDependent::SphericalCompression<
```
`const` several times

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+      bad_max_radius, missing_f_of_t);
+}
+
+bool random_bool(gsl::not_null<std::mt19937*> generator) noexcept {
```
`const gsl`

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+}
+
+template <bool InteriorMap>
+void test_out_of_bounds(gsl::not_null<std::mt19937*> generator) noexcept {
```
`const gsl`

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+
+template <bool InteriorMap>
+void test_out_of_bounds_inverse(
+    gsl::not_null<std::mt19937*> generator) noexcept {
```
`const gsl`

You're missing them everywhere in this file, I'll just stop commenting and let you find them with a search :)

nilsdeppe commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -0,0 +1,357 @@
+    const std::array<double, 3> target_radial_position{
+        radial_position(target_coords, center_)};
+    const double target_radius{magnitude(target_radial_position)};
+    check_target_radius<InteriorMap>(target_radius, min_radius_ - lambda_y,
```
Why doesn't this return a `bool` and then you return a `std::nullopt` because you couldn't invert the map? Doesn't the current design interfere with the idea that `if we can't find the inverse, that's okay?` which we need for a number of different things.

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -0,0 +1,564 @@
+                      std::numeric_limits<double>::epsilon();
+    }
+  }
+  map.inverse(point(target_radius), time, functions_of_time);
```
if the `check_target_radius` is changed to return `std::nullopt` this test will no longer need to be an error test

geoffrey4444 commented:
The test failure is a random failure in code untouched by this PR...I appended my failure to the issue for it, #1765

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_SphericalCompression.cpp:
```diff
@@ -549,16 +546,13 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.SphericalCompression",
 }
 
-// [[OutputRegex, Target radius]]
 [[noreturn]] SPECTRE_TEST_CASE(
```
while squashing, could you combine this test case with the one on line 468, maybe just sticking the code into a free function and passing in the generator? Reduces the number of tests helps keep test runtime down :)

geoffrey4444 replied:
Turns out I don't need the if logic anymore (that was for when it was an error test), so I just added both tests to the test on line 468. Since it's just two lines now, I figure no need for a free function

nilsdeppe reviewed (commented):
Looks good, please go ahead and squash, including any additional changes.

geoffrey4444 commented:
Ping @nilsdeppe @markscheel

nilsdeppe commented:
I'm not sure why I'm being pinged. I'm waiting for  a response on https://github.com/sxs-collaboration/spectre/pull/2720#discussion_r590935128

geoffrey4444 commented:
> I'm not sure why I'm being pinged. I'm waiting for a response on [#2720 (comment)](https://github.com/sxs-collaboration/spectre/pull/2720#discussion_r590935128)

Hi @nilsdeppe ...I think there was miscommunication here...sorry! I didn't realize that the comment you linked needed further comment to approve and merge this PR, because that was just discussing changes that would be implemented, maybe, in a future PR....I've actually been waiting a week to see if you approved my squash. Oops!

Anyway, I replied that you might be right, but I'm not sure, and anyway I don't think the resolution of that discussion will affect this PR. Is there something else you'd like me to do before you're willing to take another look at this PR? If so, please let me know.
